internal/pkg/policy/policy_output.go

```diff
@@ -32,163 +32,208 @@ var (
+			Str("roles", string(p.Role.Raw)).
+			Msg("Updating agent record to pick up most recent roles.")
 
-		} else if needNewKey {
-			zlog.Debug().
-				RawJSON("roles", p.Role.Raw).
-				Str("oldHash", agent.PolicyOutputPermissionsHash).
-				Str("newHash", p.Role.Sha2).
-				Msg("Generating a new API key")
-
-			outputAPIKey, err := generateOutputAPIKey(ctx, bulker, agent.Id, p.Name, p.Role.Raw)
-			if err != nil {
-				zlog.Error().Err(err).Msg("fail generate output key")
-				return err
-			}
+		fields := map[string]interface{}{
+			dl.FieldPolicyOutputPermissionsHash: p.Role.Sha2,
+		}
 
-			agent.DefaultAPIKey = outputAPIKey.Agent()
+		// Using painless script to update permission hash for updated key
+		body, err := renderUpdatePainlessScript(p.Name, fields)
+		if err != nil {
+			return err
+		}
 
-			// When a new keys is generated we need to update the Agent record,
-			// this will need to be updated when multiples Elasticsearch output
-			// are used.
-			zlog.Info().
-				Str("hash.sha256", p.Role.Sha2).
-				Str(logger.DefaultOutputAPIKeyID, outputAPIKey.ID).
-				Msg("Updating agent record to pick up default output key.")
+		if err = bulker.Update(ctx, dl.FleetAgents, agent.Id, body); err != nil {
+			zlog.Error().Err(err).Msg("fail update agent record")
+			return err
+		}
 
-			fields := map[string]interface{}{
-				dl.FieldDefaultAPIKey:               outputAPIKey.Agent(),
-				dl.FieldDefaultAPIKeyID:             outputAPIKey.ID,
-				dl.FieldPolicyOutputPermissionsHash: p.Role.Sha2,
-			}
-			if agent.DefaultAPIKeyID != "" {
-				fields[dl.FieldDefaultAPIKeyHistory] = model.DefaultAPIKeyHistoryItems{
-					ID:        agent.DefaultAPIKeyID,
-					RetiredAt: time.Now().UTC().Format(time.RFC3339),
-				}
-			}
+	} else if needNewKey {
+		zlog.Debug().
+			RawJSON("fleet.policy.roles", p.Role.Raw).
+			Str("fleet.policy.default.oldHash", output.PermissionsHash).
+			Str("fleet.policy.default.newHash", p.Role.Sha2).
+			Msg("Generating a new API key")
+
+		ctx := zlog.WithContext(ctx)
+		outputAPIKey, err :=
+			generateOutputAPIKey(ctx, bulker, agent.Id, p.Name, p.Role.Raw)
+		if err != nil {
+			return fmt.Errorf("failed generate output API key: %w", err)
+		}
 
-			// Using painless script to append the old keys to the history
-			body, err := renderUpdatePainlessScript(fields)
+		// When a new keys is generated we need to update the Agent record,
+		// this will need to be updated when multiples remote Elasticsearch output
+		// are supported.
+		zlog.Info().
+			Str("fleet.policy.role.hash.sha256", p.Role.Sha2).
+			Str(logger.DefaultOutputAPIKeyID, outputAPIKey.ID).
+			Msg("Updating agent record to pick up default output key.")
+
+		fields := map[string]interface{}{
+			dl.FieldPolicyOutputAPIKey:          outputAPIKey.Agent(),
+			dl.FieldPolicyOutputAPIKeyID:        outputAPIKey.ID,
+			dl.FieldPolicyOutputPermissionsHash: p.Role.Sha2,
+		}
 
-			if err != nil {
-				return err
+		if !foundOutput {
```

Comment:
What was the fix for the bug? I see this if statement added in the commit that contains a fix, along with the block below being moved later, but it isn't actually clear to me why this fixes the problem.

```go
		// Now that all is done, we can update the output on the agent variable
		// Right not it's more for consistency and to ensure the in-memory agent
		// data is correct and in sync with ES, so it can be safely used after
		// this method returns.
		output.Type = OutputTypeElasticsearch
		output.APIKey = outputAPIKey.Agent()
		output.APIKeyID = outputAPIKey.ID
		output.PermissionsHash = p.Role.Sha2 // for the sake of consistency
```